### ISSUE TITLE
Question: Using `ContinuousStreamingEnabled` with OEM for US images from BK5000

### ISSUE THREAD

**Issue by Roland-djee** [PlusToolkit/PlusLib#486]
Hello there,
It looks like the `ContinuousStreamingEnabled` option doesn't improve the rendering of US images on Slicer we've acquired from a BK5000, as opposed to the 3-4 frames per second we expect (and see) without. Is this something to do with the TCP/IP enabled in the OEM ? Would it make a difference to swap for the CameraLink ? 
Thank you for your help.

**Comment by adamrankin** [comment]
I am not familiar with the BK interface, @lassoan could you comment?

**Comment by lassoan** [comment]
CameraLink is an entirely different story. It provides hundreds of megabytes per second bandwidth and low-level access to all image streams (including raw RF data). However, it requires a CameraLink capture card, special cable, and access to the CameraLink bus on the ultrasound system.

Have you purchased access to the GRAB_FRAME API (it used to be a separate additional license on top of the OEM interface)?

Could you attach your Plus configuration file?

**Comment by UCLGuichard** [comment]
Here it is. For this example, we turned `ContinuousStreamingEnabled="true"` without noticing a difference. Wouldn't we be notified if we needed another licence ?

```
 <PlusConfiguration version="2.1">

  <DataCollection StartupDelaySec="1.0" >
    <DeviceSet 
      Name="PlusServer: BK ultrasound imaging device using OEM interface"
      Description="Broadcasting ultrasound images acquired from the BK system through OEM (TCP/IP) interface."
    />
    <!-- Setting AcquisitionRate to higher value may decrease the responsiveness of the BK software on the scanner. Thoroughly test the performance of the BK scanner when acquiring images through the OEM interface. -->
    <!-- For offline testing set OfflineTesting to true and supply path to test image file. -->
    <Device 
      Id="VideoDevice"
      Type="BkProFocusOem"
      ScannerAddress="128.16.0.3"
      OemPort="7915"
      ContinuousStreamingEnabled="false"
      ColorEnabled="false"
      AcquisitionRate="1.5"
      OfflineTesting="false"
      OfflineTestingFilePath="c:\dev\bktest.png" >
      <DataSources>
        <DataSource Type="Video" Id="Video" PortUsImageOrientation="MF"  />
      </DataSources>
      <OutputChannels>
        <OutputChannel Id="VideoStream" VideoDataSourceId="Video"  />
      </OutputChannels>
    </Device>
    <Device
      Id="CaptureDevice"
      Type="VirtualCapture"
      BaseFilename="RecordingTest.mha"
      EnableCapturingOnStart="FALSE" >
      <InputChannels>
        <InputChannel Id="VideoStream" />
      </InputChannels>
    </Device>
  </DataCollection>

  <CoordinateDefinitions>
    <Transform From="Image" To="Reference"
      Matrix="
        0.2 0.0 0.0 0.0
        0.0 0.2 0.0 0.0
        0.0 0.0 0.2 0.0        
        0 0 0 1" />
  </CoordinateDefinitions>
  
  <PlusOpenIGTLinkServer 
    MaxNumberOfIgtlMessagesToSend="1" 
    MaxTimeSpentWithProcessingMs="50" 
    ListeningPort="18944" 
    SendValidTransformsOnly="false" 
    OutputChannelId="VideoStream" > 
    <DefaultClientInfo> 
      <MessageTypes> 
        <Message Type="IMAGE" />
        <Message Type="STRING" />
      </MessageTypes>
      <!-- Send parameters according to OpenIGTLinkIO standard (under development) -->
      <!-- These parameters should be sent as meta data in the image message in the future. -->
      <StringNames>
        <String Name="ProbeType" />
        <String Name="Origin" />
        <String Name="Angles" />
        <String Name="BouningBox" />
        <String Name="Depths" />
        <String Name="LinearWidth" />
        <!-- Spacing are sent as separate messages, should be sent with image in the future. -->
        <String Name="SpacingX" />
        <String Name="SpacingY" />
      </StringNames>
      <ImageNames>
        <Image Name="Image" EmbeddedTransformToFrame="Reference" />
      </ImageNames>
    </DefaultClientInfo>
  </PlusOpenIGTLinkServer>

</PlusConfiguration>

```

**Comment by lassoan** [comment]
Try adjusting ContinuousStreamingEnabled and AcquisitionRate and check what difference they make. For example if ContinuousStreamingEnabled="false" and you set AcquisitionRate to more than (e.g., 15) then the ultrasound system may become very laggy (console responding very slowly). If you set AcquisitionRate to a very low value (0.2) then you should see image is updated once in every few seconds.

If you use 3D Slicer for visualization then you can use Node Modified Statistics module (in DebuggingTools extension) to measure image update rate.

**Comment by UCLGuichard** [comment]
What we checked so far:
- In ContinuousStreamingEnabled="false":
We tried different acquisition rates, from low (~1) to high (~10-20 or 30) and in the low case, we clearly see 1 fps. In the highest cases, it sort of saturates ~2-3 fps.
- In ContinuousStreamingEnabled="true":
Irrespective of the acquisition rate we set in, no changes are observable.

**Comment by lassoan** [comment]
> In ContinuousStreamingEnabled="true": Irrespective of the acquisition rate we set in, no changes are observable.

Good. This indicates that continuous streaming works. Maybe need to adjust the frame rate.

Could you collect a log file with DEBUG level logging, upload it somewhere, and post the link here?

**Comment by ageorgou** [comment]
Thanks for the suggestions @lassoan! I work with @UCLGuichard, so facing the same issue.

Concerning the license, we are told that we have the appropriate one. A different piece of software that also uses GRAB_FRAME works fine with the same machine (the license would be per-machine, correct?)

Just to clarify, when disabling continuous streaming, we are only getting ~3-4 fps, which we understand is to be expected. The problem appears when enabling continuous streaming, in which case we get (going by eye) fewer frames, and with a noticeable lag. Thanks for suggesting the Node Modified Statistics module, we’ll take a look at that for more precise information - if I understand right, that should give us the rate that Slicer receives images from Plus Server via OpenIGTLink?

As a side note, I think that setting `AcquisitionRate` doesn’t have any effect when continuous streaming is enabled, as Plus Server will always request 20 fps: https://github.com/PlusToolkit/PlusLib/blob/2503fe99eb875a5cc9bc9eac9d6b262df9b05fdb/src/PlusDataCollection/BkProFocus/vtkPlusBkProFocusOemVideoSource.cxx#L204-L220

**Comment by lassoan** [comment]
> As a side note, I think that setting AcquisitionRate doesn’t have any effect when continuous streaming is enabled, as Plus Server will always request 20 fps (we don't see any difference by changing it either)

Yes, probably the code should be changed to set the frame rate based on the acquisition rate specified in the config file (instead of hardcoded "20"). We don't have a BK machine here, so it would be great if you could make the change, test it, and if it works then send a pull request.

In streaming mode, we process all messages we receive from the machine at the frame rate that the machine sends them. Lower frame rates can be caused by anything (slow network, large image size, high CPU load, etc.), so you need check the frame rate at several places in the pipeline to see where the bottleneck is.

**Comment by ageorgou** [comment]
Thanks for confirming @lassoan. We're happy to send a pull request for configuring the acquisition rate if we get it working!

Yes, we're not sure why the apparent frame rate is so low when streaming, but we're not sure how to check the rate at different points. The Slicer module might help, or we'll try rebuilding Plus Server with more detailed timestamps (or examine the OEM logs in the BK machine, if we can access those).

**Comment by lassoan** [comment]
You get much more details in the log file if you set log level to DEBUG (you can even try TRACE level for maximum details, but that might significantly impact performance).

**Comment by UCLGuichard** [comment]
Hi and thanks,
We actually get an improvement in the Slicer rendering of the US images. From the statistical tool you provide us with, we observe that we reach a limit ~13fps when the PLUS server is built with GRAB_FRAME=20, 30 or 40. Maybe any thoughts to push that a bit higher ?

**Comment by lassoan** [comment]
What is the frame rate on the ultrasound system? Can you get higher frame rate with the other software that you mentioned?

13fps might be the maximum what the OEM interface can provide. You might get higher frame rates by changing settings on the BK scanner (changing imaging parameters, image size, compression settings, etc. - I'm just guessing, I don't know what actual settings are available).

13fps may be good enough for most interventions and as far as I remember this was about what was achieved by others who used BK's OEM interface. For example Raj Shekhar's team at Children's (see [paper](https://link.springer.com/article/10.1007/s00464-014-3433-x)); or more recently by @olevs at SINTEF.

Note that for good performance low latency is very important, too.

**Comment by UCLGuichard** [comment]
Well, there is scarcely any mention of the fps in the KB manual unfortunately.

Yes, with the other software we reached ~40fps. But this is without passing through the OEM, PLUS and the OpenIGTLink though.

My suspicion is that the OEM is the bottleneck though the least controllable.

**Comment by lassoan** [comment]
What interface did you use with the other software? Just a framegrabber? What is the FPS on the ultrasound system?

**Comment by ageorgou** [comment]
Just to clarify, the other software does use OEM (GRAB_FRAME), requesting 40fps. It also reports a frame rate of ~40.

**Comment by UCLGuichard** [comment]
That's right thanks

**Comment by lassoan** [comment]
Does it request 40fps the same way as Plus? Does it actually receive frames at 40FPS or is that rendering update rate?

If you do not show the received images in Slicer (not select it as background image in Slice viewers, not use it in volume reslice driver) does the image node update rate gets higher?

What operating system Slicer is running on and what Slicer version do you use?

**Comment by olevs** [comment]
To test the fps of your pipeline without actually connecting to the BK machine you can set OfflineTesting to true, and have OfflineTestingFilePath pointing to a test image to send.

The hardcoded 20 fps is probably my fault. Seems like the current implementation don't use the AcquisitionRate variable for setting the frame rate (I see it's not even reading it from the xml file). I didn't have time to implement more features.

When everything was working ok, I think we always got 20 fps (on Ubuntu 16.04).

**Comment by ageorgou** [comment]
Thanks @olevs!

> To test the fps of your pipeline without actually connecting to the BK machine you can set OfflineTesting to true, and have OfflineTestingFilePath pointing to a test image to send.
We'll try that, we were wondering about the OfflineTesting too.

As for the AcquisitionRate, while it's not used to set the machine's rate, in some of our attempts it did seem to have some effect on the final result (e.g. by changing it to 1). We think that's because it's used in the the thread that reads the incoming messages, to decide how often it should check for a new response. That said, we still haven't figured out where its value is read from the XML config...

**Comment by ageorgou** [comment]
Ah yes, tracked it down in the `XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_READING` macro! I see where the acquisition rate is read now.

**Comment by rc-softdev-admin** [comment]
@lassoan 
Few clarifications: the other software will send a message to the scanner to start continuous streaming at 40fps. For each image received, it is saved, and the data source keeps track of the number of frames per second. Then a separate gui thread ticks over at a completely independent rate. So you can set the screen to refresh say at 5-20Hz, while data is collected independently. Default screen update is 20fps.
If that makes sense

**Comment by UCLGuichard** [comment]
Sorry I got confused between accounts. I actually wrote the message above.

**Comment by UCLGuichard** [comment]
Hi @lassoan 
Thank you for the reply and sorry for my late one.
We've got now some good incoming ~50fps to PLUS from the BK scanner. However, we'd like now to use a complementary StealthStation, meaning using the Stealthlink. So far, our build of PLUS systematically fails on a Windows machine with VS2010, Qt5.1.1 and WIN32 when building VTK. I have read in #83 that you successfully built PLUS on 64-bit platforms. Could you please tell me a bit more about it ? That'd be greatly helpful.

**Comment by ageorgou** [comment]
An overdue update:

We have tried to build the PlusServer on different platforms (OS X and Windows), so that the acquisition rate (in streaming mode) is read from the configuration file rather than set to 20fps. We have had mixed results: on one Mac laptop, the frame rate has been matching the config setting, and we have been getting frame rates up to 50fps (according to the Plus diagonstics, and with only a small drop in Slicer). However, in other laptops (one Mac and two Windows), we have not managed to get more than 15fps (according to the Plus diagnostics). Oddly, this has been happening even in offline testing mode. We have not yet looked into why that is in detail.

@lassoan Does it make sense to send a PR with the changes (reading the AcquisitionRate from the XML config in continuous streaming mode)? We have only got it working in one machine, but then we could not get 20fps on the other machines before the changes either...
